src/dvc_data/index/diff.py

```diff
@@ -102,13 +108,20 @@ def _diff_entry(
     if hash_only:
         return hi_diff
 
-    if meta_diff != UNCHANGED:
+    # If both meta's are None, return hi_diff
+    if meta_diff == UNCHANGED and old_meta is None:
+        return hi_diff
+
+    # If both hi's are falsey, return meta_diff
+    if hi_diff == UNCHANGED and not old_hi:
         return meta_diff
 
-    if hi_diff != UNCHANGED:
-        return hi_diff
```

Comment:
This only worked because previously `meta_diff` could only be `UNCHANGED` or `MODIFY`, but we should really be able to return `ADD`/`DELETE` when we are diffing with `meta_only=True`